File: ncPersistencia/pClienteEndereco.vb
Imports ncDados.nsCliente
Imports ncComum.nsAcessoBD
Imports ncComum.nsFuncoes
Imports ncComum.nsExcecao

Namespace nsCliente

  Public Class pClienteEndereco

    Public Function Listar() As ColecaoClienteEndereco

      Dim retorno As ColecaoClienteEndereco
      Dim acessoBanco As cAcessoBD
      Dim ds As DataSet
      Dim dt As DataTable
      Dim row As DataRow
      Dim item As dClienteEndereco
      Dim comandoSQL As String

      Try

        acessoBanco = New cAcessoBD

        comandoSQL = " Select logradouro, numero, complemento, cidade, estado_cid, cep, " & _
            " dataInclusao, tipoResidencia, bairro, tipoEndereco, cliente_cid From clienteenderecos "

        ds = acessoBanco.ExecutarDS(comandoSQL)

        If Not ds Is Nothing Then
          If ds.Tables.Count > 0 Then
            dt = ds.Tables(0)

            If dt.Rows.Count > 0 Then
              retorno = New ColecaoClienteEndereco

              For Each row In dt.Rows
                item = New dClienteEndereco

                item.logradouro = cFuncoes.RetornarTexto(row("logradouro"))
                item.numero = cFuncoes.RetornarInteiro(row("numero"))
                item.complemento = cFuncoes.RetornarTexto(row("complemento"))
                item.cidade = cFuncoes.RetornarTexto(row("cidade"))
                item.estado_cid = cFuncoes.RetornarInteiro(row("estado_cid"))
                item.cep = cFuncoes.RetornarInteiro(row("cep"))
                item.dataInclusao = cFuncoes.RetornarTexto(row("dataInclusao"))
                item.tipoResidencia = cFuncoes.RetornarTexto(row("tipoResidencia"))
                item.bairro = cFuncoes.RetornarTexto(row("bairro"))
                item.tipoEndereco = cFuncoes.RetornarTexto(row("tipoEndereco"))
                item.cliente_cid = cFuncoes.RetornarInteiro(row("cliente_cid"))

                retorno.Add(item)
              Next
            Else
              retorno = Nothing
            End If
          Else
            retorno = Nothing
          End If
        Else
          retorno = Nothing
        End If

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Listar Cliente - Endereço [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Listar = retorno

    End Function

    Public Function Consultar(ByVal dados As dClienteEndereco) As ColecaoClienteEndereco

      Dim retorno As ColecaoClienteEndereco
      Dim acessoBanco As cAcessoBD
      Dim ds As DataSet
      Dim dt As DataTable
      Dim row As DataRow
      Dim item As dClienteEndereco
      Dim sqlSelect As String
      Dim sqlWhere As String
      Dim sqlFrom As String

      Try

        acessoBanco = New cAcessoBD


        sqlSelect = " SELECT c.logradouro, c.numero, c.complemento, c.cidade, c.estado_cid, c.cep, " & _
            " c.dataInclusao, c.tipoResidencia, c.bairro, c.tipoEndereco, c.cliente_cid, e.sigla "
        sqlWhere = String.Empty
        sqlFrom = " From clienteenderecos c LEFT OUTER JOIN estados e ON c.estado_cid = e.cid "

        '-- logradouro
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.logradouro, "c.logradouro")

        '-- numero
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.numero, "c.numero")

        '-- complemento
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.complemento, "c.complemento")

        '-- cidade
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.cidade, "c.cidade")

        '-- estado_cid
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.estado_cid, "c.estado_cid")

        '-- cep
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.cep, "c.cep")

        '-- dataInclusao
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.dataInclusao, "c.dataInclusao")

        '-- tipoResidencia
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.tipoResidencia, "c.tipoResidencia")

        '-- bairro
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.bairro, "c.bairro")

        '-- tipoEndereco
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.tipoEndereco, "c.tipoEndereco")

        '-- cliente_cid
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.cliente_cid, "c.cliente_cid")

        If Not sqlWhere.Equals(String.Empty) Then
          sqlWhere = " WHERE " & sqlWhere
        End If

        ds = acessoBanco.ExecutarDS(sqlSelect & " " & sqlFrom & " " & sqlWhere)

        If Not ds Is Nothing Then
          If ds.Tables.Count > 0 Then
            dt = ds.Tables(0)

            If dt.Rows.Count > 0 Then
              retorno = New ColecaoClienteEndereco

              For Each row In dt.Rows
                item = New dClienteEndereco

                item.logradouro = cFuncoes.RetornarTexto(row("logradouro"))
                item.numero = cFuncoes.RetornarInteiro(row("numero"))
                item.complemento = cFuncoes.RetornarTexto(row("complemento"))
                item.cidade = cFuncoes.RetornarTexto(row("cidade"))
                item.estado_cid = cFuncoes.RetornarInteiro(row("estado_cid"))
                item.cep = cFuncoes.RetornarInteiro(row("cep"))
                item.dataInclusao = cFuncoes.RetornarTexto(row("dataInclusao"))
                item.tipoResidencia = cFuncoes.RetornarTexto(row("tipoResidencia"))
                item.bairro = cFuncoes.RetornarTexto(row("bairro"))
                item.tipoEndereco = cFuncoes.RetornarTexto(row("tipoEndereco"))
                item.cliente_cid = cFuncoes.RetornarInteiro(row("cliente_cid"))
                item.siglaEstado = cFuncoes.RetornarTexto(row("sigla"))

                retorno.Add(item)
              Next
            Else
              retorno = Nothing
            End If
          Else
            retorno = Nothing
          End If
        Else
          retorno = Nothing
        End If

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Consultar Cliente - Endereço [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Consultar = retorno

    End Function

    Public Function Incluir(ByVal dados As dClienteEndereco) As Integer

      Dim retorno As Integer
      Dim acessoBanco As cAcessoBD
      Dim comandoSQL As String

      Try

        acessoBanco = New cAcessoBD

        comandoSQL = " INSERT INTO " & _
            " clienteenderecos (logradouro, numero, complemento, cidade, estado_cid, cep, " & _
            " dataInclusao, tipoResidencia, bairro, tipoEndereco, cliente_cid) " & _
            " VALUES (" & _
            cFuncoes.PersistirTexto(dados.logradouro) & "," & _
            cFuncoes.PersistirInteiro(dados.numero) & "," & _
            cFuncoes.PersistirTexto(dados.complemento) & "," & _
            cFuncoes.PersistirTexto(dados.cidade) & "," & _
            cFuncoes.PersistirInteiro(dados.estado_cid) & "," & _
            cFuncoes.PersistirInteiro(dados.cep) & "," & _
            cFuncoes.PersistirTexto(dados.dataInclusao) & "," & _
            cFuncoes.PersistirTexto(dados.tipoResidencia) & "," & _
            cFuncoes.PersistirTexto(dados.bairro) & "," & _
            cFuncoes.PersistirTexto(dados.tipoEndereco) & "," & _
            cFuncoes.PersistirInteiro(dados.cliente_cid) & ")"

        retorno = acessoBanco.ExecutarCID(comandoSQL)

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Incluir Cliente - Endereço [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Incluir = retorno

    End Function

    Public Function ExcluirPorCliente(ByVal cliente_cid As Integer) As Integer

      Dim retorno As Integer
      Dim acessoBanco As cAcessoBD
      Dim comandoSQL As String

      Try

        acessoBanco = New cAcessoBD

        comandoSQL = " DELETE FROM clienteenderecos " & _
            " WHERE cliente_cid = " & cliente_cid.ToString()

        retorno = acessoBanco.ExecutarINT(comandoSQL)

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Excluir Cliente - Endereço [" & Me.ToString() & "] - " & ex.Message)

      End Try

      ExcluirPorCliente = retorno

    End Function

  End Class

End Namespace

File: ncRepositorios/pClienteEndereco.cs
using System;
using System.Collections.Generic;
using Dapper;
using MySql.Data.MySqlClient;
using ncNComum;
using ncModelos;

namespace ncRepositorios
{
    public class pClienteEndereco : RepositorioBase, IpClienteEndereco
    {
        public ColecaoClienteEndereco Listar()
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    var lista = conn.Query<dClienteEndereco>(
                        "SELECT logradouro, numero, complemento, cidade, estado_cid, cep, " +
                        "dataInclusao, tipoResidencia, bairro, tipoEndereco, cliente_cid FROM clienteenderecos").AsList();
                    if (lista.Count == 0) return null;
                    var retorno = new ColecaoClienteEndereco();
                    retorno.AddRange(lista);
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Listar Cliente - Endereço [" + ToString() + "] - " + ex.Message);
            }
        }

        public ColecaoClienteEndereco Consultar(dClienteEndereco dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    var conditions = new List<string>();
                    var p = new DynamicParameters();
                    if (!string.IsNullOrEmpty(dados.logradouro)) { conditions.Add("c.logradouro=@logradouro"); p.Add("logradouro", dados.logradouro); }
                    if (dados.numero != null && dados.numero != 0) { conditions.Add("c.numero=@numero"); p.Add("numero", dados.numero); }
                    if (!string.IsNullOrEmpty(dados.complemento)) { conditions.Add("c.complemento=@complemento"); p.Add("complemento", dados.complemento); }
                    if (!string.IsNullOrEmpty(dados.cidade)) { conditions.Add("c.cidade=@cidade"); p.Add("cidade", dados.cidade); }
                    if (dados.estado_cid != null && dados.estado_cid != 0) { conditions.Add("c.estado_cid=@estado_cid"); p.Add("estado_cid", dados.estado_cid); }
                    if (dados.cep != null && dados.cep != 0) { conditions.Add("c.cep=@cep"); p.Add("cep", dados.cep); }
                    if (!string.IsNullOrEmpty(dados.dataInclusao)) { conditions.Add("c.dataInclusao=@dataInclusao"); p.Add("dataInclusao", dados.dataInclusao); }
                    if (!string.IsNullOrEmpty(dados.tipoResidencia)) { conditions.Add("c.tipoResidencia=@tipoResidencia"); p.Add("tipoResidencia", dados.tipoResidencia); }
                    if (!string.IsNullOrEmpty(dados.bairro)) { conditions.Add("c.bairro=@bairro"); p.Add("bairro", dados.bairro); }
                    if (!string.IsNullOrEmpty(dados.tipoEndereco)) { conditions.Add("c.tipoEndereco=@tipoEndereco"); p.Add("tipoEndereco", dados.tipoEndereco); }
                    if (dados.cliente_cid != null && dados.cliente_cid != 0) { conditions.Add("c.cliente_cid=@cliente_cid"); p.Add("cliente_cid", dados.cliente_cid); }
                    var where = conditions.Count > 0 ? "WHERE " + string.Join(" AND ", conditions) : "";
                    var sql = $"SELECT c.logradouro, c.numero, c.complemento, c.cidade, c.estado_cid, c.cep, " +
                              $"c.dataInclusao, c.tipoResidencia, c.bairro, c.tipoEndereco, c.cliente_cid, e.sigla AS siglaEstado " +
                              $"FROM clienteenderecos c LEFT OUTER JOIN estados e ON c.estado_cid = e.cid {where}";
                    var lista = conn.Query<dClienteEndereco>(sql, p).AsList();
                    if (lista.Count == 0) return null;
                    var retorno = new ColecaoClienteEndereco();
                    retorno.AddRange(lista);
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Consultar Cliente - Endereço [" + ToString() + "] - " + ex.Message);
            }
        }

        public int Incluir(dClienteEndereco dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    conn.Execute(
                        "INSERT INTO clienteenderecos (logradouro, numero, complemento, cidade, estado_cid, cep, " +
                        "dataInclusao, tipoResidencia, bairro, tipoEndereco, cliente_cid) " +
                        "VALUES (@logradouro, @numero, @complemento, @cidade, @estado_cid, @cep, " +
                        "@dataInclusao, @tipoResidencia, @bairro, @tipoEndereco, @cliente_cid)",
                        new {
                            logradouro = dados.logradouro, numero = dados.numero, complemento = dados.complemento,
                            cidade = dados.cidade, estado_cid = dados.estado_cid, cep = dados.cep,
                            dataInclusao = dados.dataInclusao, tipoResidencia = dados.tipoResidencia,
                            bairro = dados.bairro, tipoEndereco = dados.tipoEndereco, cliente_cid = dados.cliente_cid
                        });
                    return (int)conn.ExecuteScalar<long>("SELECT LAST_INSERT_ID()");
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Incluir Cliente - Endereço [" + ToString() + "] - " + ex.Message);
            }
        }

        public int ExcluirPorCliente(int cliente_cid)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    return conn.Execute("DELETE FROM clienteenderecos WHERE cliente_cid=@cliente_cid", new { cliente_cid });
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Excluir Cliente - Endereço [" + ToString() + "] - " + ex.Message);
            }
        }
    }
}
